src/bridge/cockpitrouter.c

```diff
@@ -323,6 +365,104 @@ process_open_peer (CockpitRouter *self,
   return cockpit_peer_handle (peer, channel, options, data);
 }
 
+static GBytes *
+substitute_json_string (const gchar *variable,
+                        gpointer user_data)
+{
+  const gchar *value;
+  JsonObject *options = user_data;
+  if (options && cockpit_json_get_string (options, variable, "", &value))
+    return g_bytes_new (value, strlen (value));
+  else if (options)
+    g_message ("Couldn't get argument for bridge: got invalid value for '%s'", variable);
+
+  return g_bytes_new ("", 0);
+}
+
+static JsonArray *
+build_dynamic_array (gchar **config_args,
+                     JsonObject *options)
+{
+  gint length;
+  gint i;
+  JsonArray *args;
+
+  g_return_val_if_fail (config_args != NULL, NULL);
+  args = json_array_new ();
+
+  length = g_strv_length (config_args);
+  for (i = 0; i < length; i++)
+    {
+      GByteArray *array = g_byte_array_new ();
+      GBytes *input = g_bytes_new_static (config_args[i], strlen(config_args[i]) + 1);
+      GList *output = cockpit_template_expand (input, substitute_json_string,
+                                               "${", "}", options);
+      GList *l;
+      for (l = output; l != NULL; l = g_list_next (l))
+        {
+          gsize size;
+          gpointer data = g_bytes_unref_to_data (l->data, &size);
+          g_byte_array_append (array, data, size);
+        }
+
+      json_array_add_string_element (args, (const gchar *)g_byte_array_free (array, FALSE));
+
+      g_bytes_unref (input);
+      g_list_free (output);
+    }
+
+  return args;
+}
+
+static gboolean
+process_open_dynamic_peer (CockpitRouter *self,
+                           const gchar *channel,
+                           JsonObject *options,
+                           GBytes *data,
+                           gpointer user_data)
+{
+  CockpitPeer *peer = NULL;
+  DynamicPeer *dp = user_data;
+  GBytes *bytes = NULL;
+  JsonObject *config = json_object_new ();
+
+  if (json_object_has_member (dp->config, "problem"))
+    {
+      json_object_set_member (config, "problem",
+                              json_object_get_member (dp->config, "problem"));
+    }
+
+  if (json_object_has_member (dp->config, "directory"))
+    {
+      json_object_set_member (config, "directory", json_object_get_member (dp->config, "directory"));
+    }
+
+  if (dp->spawn)
+    {
+      json_object_set_array_member (config, "spawn",
+                                    build_dynamic_array (dp->spawn, options));
+    }
+
+  if (dp->env)
+    {
+      json_object_set_array_member (config, "environ",
+                                    build_dynamic_array (dp->env, options));
+    }
+
+  bytes = cockpit_json_write_bytes (config);
```

Comment:
Is this deterministic? I would love for it to be so simple, but I really suspect it's not. Very likely the initial implementation will need to use a O(N) ```GList``` and ```cockpit_json_equal()``` and then a follow up would need to implement a new ```cockpit_json_hash()``` function in a ```GHashTable```.